### PR TITLE
Support pulling OCI images

### LINKED ISSUE #37: Add OCI Distribution support
We should be able to pull `oci://example.com/foo/bar:latest` into Wagi.

### PR DESCRIPTION
**Pull request by technosophos**
Adds a mapping from the scheme `oci:` to an OCI Distributions registry.

Closes #37

Signed-off-by: Matt Butcher <matt.butcher@microsoft.com>

### CODE REVIEW

File: docs/configuring_and_running.md [hunk 1]
```diff
@@ -60,17 +60,23 @@ Each `[[module]]` section in the `modules.toml` file is responsible for mapping
...
-Relative paths will be resolved from the current working directory in which `wagi` was started.
+### Module References
+
+A module reference is a URL. There are three supported module reference schemes:
+
+- `file://`: A path to a `.wasm` or `.wat` file on the filesystem. We recommend using absolute paths beginning with `file://`. Right now, there is legacy support for absolute and relative paths without the `file://` prefix, but we discourge using that. Relative paths will be resolved from the current working directory in which `wagi` was started.
+- `bindle:`: A reference to a Bindle. This will be looked up in the configured Bindle server. Example: `bindle:example.com/foo/bar/1.2.3`.
+- `oci`: A reference to an OCI image in an OCI registry. Example: `oci:foo/bar:1.2.3` (equivalent to the Docker image `foo/bar:1.2.3`).
+
+> Note: Bindle URLs should not ever have a `//`. OCI URLs should use a `//` only if refering to a specific host, and may always omit `//` even if there is a reference to an external host.
```
Comment (radu-matei) [inline docs/configuring_and_running.md:79]: typo: _referring_

File: src/runtime/mod.rs [hunk 2]
```diff
@@ -579,11 +589,40 @@ impl Module {
...
+        let auth = RegistryAuth::Anonymous;
+
+
+        let img = url_to_oci(uri)?;
+        let data = oc.pull(&img, &auth, vec!["application/wasm"]).await?;
+        if data.layers.is_empty() {
+            anyhow::bail!("image has no layers");
+        }
+        let first_layer = data.layers.get(0).unwrap();
+        let module = wasmtime::Module::new(&Engine::default(), first_layer.data.as_slice())?;
```
Comment (radu-matei) [inline src/runtime/mod.rs:604]: This should not create a new engine, rather use the store's engine: 

```
Error: Not all routes could be built: cross-`Engine` instantiation is not currently supported
```
Specifically:

```
                "oci" => self.load_oci(&uri, &store).await,
```

File: src/runtime/mod.rs [hunk 3]
```diff
@@ -579,11 +589,40 @@ impl Module {
...
         .await
     }
+    async fn load_oci(&self, uri: &Url) -> anyhow::Result<wasmtime::Module> {
+        let config = ClientConfig::default();
+        let mut oc = Client::new(config);
+        let auth = RegistryAuth::Anonymous;
+
+
+        let img = url_to_oci(uri)?;
+        let data = oc.pull(&img, &auth, vec!["application/wasm"]).await?;
```
Comment (radu-matei) [inline src/runtime/mod.rs:599]: This should be `application/vnd.wasm.content.layer.v1+wasm`.

File: src/runtime/mod.rs (hunk 2 above)
Comment (radu-matei) [reply src/runtime/mod.rs:604]: This also applies to the Bindle support.

File: docs/configuring_and_running.md [hunk 4]
```diff
@@ -60,17 +60,23 @@ Each `[[module]]` section in the `modules.toml` file is responsible for mapping
...
-The `module` directive is a path to a `wasm` or `wat` file on the filesystem.
-We recommend using absolute paths.
-Relative paths will be resolved from the current working directory in which `wagi` was started.
+### Module References
+
+A module reference is a URL. There are three supported module reference schemes:
+
+- `file://`: A path to a `.wasm` or `.wat` file on the filesystem. We recommend using absolute paths beginning with `file://`. Right now, there is legacy support for absolute and relative paths without the `file://` prefix, but we discourge using that. Relative paths will be resolved from the current working directory in which `wagi` was started.
+- `bindle:`: A reference to a Bindle. This will be looked up in the configured Bindle server. Example: `bindle:example.com/foo/bar/1.2.3`.
+- `oci`: A reference to an OCI image in an OCI registry. Example: `oci:foo/bar:1.2.3` (equivalent to the Docker image `foo/bar:1.2.3`).
```
Comment (radu-matei) [inline docs/configuring_and_running.md:77]: It is a bit unclear if modules should be referred to as `oci:<>` or `oci://`.

**Comment by radu-matei** [comment]
When building the routes, the module is pulled in memory and instantiated.
However, when executing, `run_wasm` expects to load it from a local file - so I assume when the module is first pulled, it should be either stored on the filesystem, or pulled every time (not ideal).

```
let module = wasmtime::Module::from_file(store.engine(), self.module.as_str())?;
```

File: docs/configuring_and_running.md (hunk 1 above)
Comment (technosophos) [reply docs/configuring_and_running.md:79]: Fixed

File: docs/configuring_and_running.md (hunk 4 above)
Comment (technosophos) [reply docs/configuring_and_running.md:77]: The technical details is that an `oci:` URL that includes a host _can_ use the `//`, but doesn't need to. I think I will just simplify that down to "you should not use //`. Unfortunately, the URL parser has special case logic for `file` that requires `//`.

File: src/runtime/mod.rs (hunk 3 above)
Comment (technosophos) [reply src/runtime/mod.rs:599]: Oh! Right! I forgot that there were special types for layers. I should probably refactor this into a const, too.

File: src/runtime/mod.rs (hunk 2 above)
Comment (technosophos) [reply src/runtime/mod.rs]: Oh, of course. 🤦  Fixed both.

**Comment by technosophos** [comment]
Okay, @radu-matei -- I think this is ready for re-review.

To test:

- Run a local OCI registry (`docker run -d -p 5000:5000 --restart always --name registry registry:2`)
- Use wasm-to-oci to push a module into that registry (`wasm-to-oci push hello.gr.wasm localhost:5000/hello-grain:1`)
- Add an entry to your `modules.toml` that looks something like this:

```toml
# OCI
[[module]]
route = "/hello-oci"
module = "oci://localhost:5000/hello-grain:1"
```

From there, you should be able to start Wagi and `curl` it.
```console
$ curl localhost:3000/hello-oci
```

**Comment by technosophos** [comment]
Also, I should clarify that I plan on going back and re-doing the Bindle code. Bindle has its own caching layer (or, rather, Bindle is supposed to completely remove the "where the heck is this located?" problem from user-space). I just need to go make sure that I wired it all up correctly. Then I think I will remove or restructure the Bindle caching layer from here.

**Review by radu-matei — APPROVED** [review]
This is awesome!
LGTM

**Pull request merged**